File: apps/website/src/lib/editor/theme.ts
import themeData from '@typie/editor/theme.json' with { type: 'json' };
import type { DarkVariant, EffectiveTheme, LightVariant } from '@typie/ui/context';

export type ThemeColors = {
  colors: Map<string, number>;
};

export type ThemeVariant = `light-${LightVariant}` | `dark-${DarkVariant}`;

const colorToU32 = (color: string, alpha = 0xff): number => {
  const clean = color.replace('#', '').replace(/^var\(.+\)$/, '');
  if (clean.length < 6) {
    console.warn(`Invalid color format: ${color}, falling back to black`);
    return 0x00_00_00_ff;
  }
  const r = Number.parseInt(clean.slice(0, 2), 16);
  const g = Number.parseInt(clean.slice(2, 4), 16);
  const b = Number.parseInt(clean.slice(4, 6), 16);
  return ((r << 24) | (g << 16) | (b << 8) | alpha) >>> 0;
};

const assembleVariant = (variant: string): Record<string, string> => {
  const isLight = variant.startsWith('light-');
  return {
    ...themeData.shared,
    ...(isLight ? themeData.lightShared : themeData.darkShared),
    ...themeData.variants[variant as keyof typeof themeData.variants],
  };
};

export const THEME_COLORS: Record<ThemeVariant, Record<string, string>> = Object.fromEntries(
  Object.keys(themeData.variants).map((variant) => [variant, assembleVariant(variant)]),
) as Record<ThemeVariant, Record<string, string>>;

const buildTheme = (rawColors: Record<string, string>): ThemeColors => ({
  colors: new Map(Object.entries(rawColors).map(([k, v]) => [k, colorToU32(v)] as [string, number])),
});

export const getEditorTheme = (effectiveTheme: EffectiveTheme, lightVariant: LightVariant, darkVariant: DarkVariant): ThemeColors => {
  const variant: ThemeVariant = effectiveTheme === 'light' ? `light-${lightVariant}` : `dark-${darkVariant}`;
  return buildTheme(THEME_COLORS[variant]);
};

File: apps/mobile/lib/screens/native_editor/state/theme.dart
import 'dart:convert';

import 'package:flutter/services.dart';

int _colorToU32(String color, [int alpha = 0xff]) {
  final clean = color.replaceAll('#', '');
  if (clean.length < 6) {
    return 0x000000ff;
  }
  final r = int.parse(clean.substring(0, 2), radix: 16);
  final g = int.parse(clean.substring(2, 4), radix: 16);
  final b = int.parse(clean.substring(4, 6), radix: 16);
  return ((r << 24) | (g << 16) | (b << 8) | alpha) & 0xffffffff;
}

Map<String, String> _assembleVariant(Map<String, dynamic> data, String variant) {
  final shared = (data['shared'] as Map<String, dynamic>).cast<String, String>();
  final isLight = variant.startsWith('light-');
  final modeShared = (data[isLight ? 'lightShared' : 'darkShared'] as Map<String, dynamic>).cast<String, String>();
  final unique = ((data['variants'] as Map<String, dynamic>)[variant] as Map<String, dynamic>).cast<String, String>();
  return {...shared, ...modeShared, ...unique};
}

Map<String, int> _buildTheme(Map<String, String> rawColors) {
  return rawColors.map((k, v) => MapEntry(k, _colorToU32(v)));
}

late Map<String, String> _lightRawColors;
late Map<String, String> _darkRawColors;
late Map<String, int> lightTheme;
late Map<String, int> darkTheme;

Future<void> initEditorTheme() async {
  final raw = await rootBundle.loadString('assets/native/theme.json');
  final data = jsonDecode(raw) as Map<String, dynamic>;

  _lightRawColors = _assembleVariant(data, 'light-white');
  _darkRawColors = _assembleVariant(data, 'dark-black');
  lightTheme = _buildTheme(_lightRawColors);
  darkTheme = _buildTheme(_darkRawColors);
}

Map<String, int> getEditorTheme(Brightness brightness) {
  return brightness == Brightness.dark ? darkTheme : lightTheme;
}

Color getEditorColor(Brightness brightness, String key) {
  final hex = (brightness == Brightness.dark ? _darkRawColors : _lightRawColors)[key]!;
  final value = int.parse(hex.substring(1), radix: 16);
  return Color(0xFF000000 | value);
}
